# $$\textbf{Python Group Project}$$

$$\text{October, 10 2024}$$

### $\text{Project Number: Project 6}$

### $\text{Project Title: Hexadecimal and Decimal Digits}$

### $\text{Group Members}$
- Chukwuezi TOCHUKWU
- Damas NIYONKURU
- Linda INGABIRE
- Darix Samani Siewe

### $\textbf{Objective}$
Write two functions, $\textbf{hex2int}$ and $\textbf{int2hex}$, that converts between hexadecimal digits ($1$,$2$,$3$,$4$,$5$,$6$,$7$,$8$,$9$,$A$,$B$,$C$,$D$,$E$,$F$) and base $10$ integers. The $\textbf{hex2int}$ function is respondible for converting a string containing several hexadecimal digits to a base 10 integer, while the $\textbf{int2hex}$ function is responsible for converting an integer between $0$ and $15$ to a single hexadecimal digit. Each function will take the value to convert as its only parameter and return the converted value as the function's only result. Ensure that the $\textbf{hex2int}$ function works correctly for both uppercase and lowercase letters. Your function should end the program with a meaningful error message if an invalid parameter is provided.

### $\textbf{Background 1}$

Given a hexadecimal number ${H = d_{l-1}d_{l-2}d_{l-3} \cdots d_2d_1d_0}$, we can find an $N$ in base $10$ such that $N = \sum_{k=0}^{l-1}16^{k}d_k$

## $\textbf{Step 1.1: Importing Relevant Modules}$

We import time and decimal modules which helps us to determine the time precision taking by our code to run.

- The `time` module helps us to determing the time taken for our code to run, helping us analyze the time complexity.
- The `decimal` module helps us determine more accurate decimal precisions especially for very small time periods.

In [2]:
import time, decimal
decimal.getcontext().prec = 10 # set the precision of decimal numbers such as time.

## $\textbf{Step 1.2: Defining the Dictionary Maps}$

 Define the dictionaries that serves as the reference for our conversion from hexadecimal to integer.

In [20]:
dic_hex2int = {"0":0, "1":1, "2":2, "3":3, "4":4, "5":5, "6":6, "7":7, "8":8, "9":9, "a":10, "b":11, "c":12, "d":13, "e":14, "f":15} # hexadecimal to integer.
print(dic_hex2int)

{'0': 0, '1': 1, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8, '9': 9, 'a': 10, 'b': 11, 'c': 12, 'd': 13, 'e': 14, 'f': 15}


## $\textbf{Step 1.3: Implementation of Algorithm for HEX to INT conversion}$

In [25]:
# Collect user input
while True: # The while loop keeps the user stuck in the loop until a valid hexadecimal number is encountered.
    hexa = input("Enter a hexadecimal number: ").lower() # converts in user's input to lower case on entry.
        
    if all([char in dic_hex2int for char in hexa]): # Checks if all digit entered is in our dictionary domain.
        break
    else:
        print("Invalid hexadecimal number. Please enter a valid hexadecimal number.\n")


#Define the dexademimal function
def hex2int(hexa):
    """
    This function takes a hexadecimal number (string) and converts it into the corresponding base 10 integer.
    
    Input Parameters:
    hexa (str) the value of the hexadecimal digit to convert.
    
    Output Parameters:
    INT_value (int) The base 10 integer value corresponding to the hexadecimal number.
    time_taken (float) The time taken to compute the conversion from hex to base 10.
    """
    # Declare a time variable to track start of algorithm implementation
    start = decimal.Decimal(time.perf_counter())
    
    l = len(hexa) # obtain the length of the input hexadecimal string.

    INT_value = 0 # this variable holds the result of the algorithm
    
    for i, digit in enumerate(hexa):
        INT_value += dic_hex2int[digit]*16**(l-(i+1)) # each digit in hexadecimal is multipled with 16 to the right power.

    # Declare a time variable to track end of algorithm implementation
    end = decimal.Decimal(time.perf_counter())
    time_taken = round((end - start), 10)
    
    print("\t")
    
    # return result and time taken to run implement the algorithm.
    return INT_value, time_taken




# Call the function and print the result and time taken to compute it.
int_value, runtime_h2i = hex2int(hexa)

# print(hex2int.__doc__)
print(f"The equivalent base_10 number is {int_value}")
print(f"The time taken is {round(runtime_h2i*10**6, 2)} micro seconds.")

Enter a hexadecimal number: 5a6f
	
The equivalent base_10 number is 23151
The time taken is 11.70 micro seconds.


### $\textbf{Background 2}$

Given an integer $N$, we can find the equivalent base_16 number by using the `Repeated Division by Base Algorithm` which returns the remainders for the division of two integers.

For instance, lets consider that $N = 23151$,

Let $H$ be the equivalent base_16 number.

To convert to base_16, we follow the iterative division as follows:

| `16` | `23151` | $$\text{Remainders}$$|
|:-----|:---------|------:|
|`16`   | `1446` |  `15`
|`16`   | `90` |  `6`|
|`16`   | `5` |  `10`|
|`16`   | `0` |  `5`|

We then collect the results in reverse order:

$$5 \rightarrow 10 \rightarrow 6 \rightarrow 15$$

We then map the remainders to equivalent base_16 values using a predefined integer-to-hexadecimal dictionary D as follows:
    $$D(5) \rightarrow D(10) \rightarrow D(6) \rightarrow D(15)$$
    
This gives:
        
$$"5" \rightarrow "a" \rightarrow "6" \rightarrow "f"$$
    
Then we join this individual strings into one string giving:
    
$$\text{"5a6f"}$$
    

## $\textbf{Step 2.2: Defining the Dictionary Maps}$

 Define the dictionaries that serves as the reference for our conversion from integer to hexadecimal.

In [7]:
dic_int2hex = {integer:hexa for hexa, integer in dic_hex2int.items()}

print(dic_int2hex)
# dic_int2hex: {0:"0", 1:"1", 2:"2", 3:"3", 4:"4", 5:"5", 6:"6", 7:"7", 8:"8", 9:"9", 10:"a", 11:"b", 12:"c", 13:"d", 14:"e", 15:"f"} # integer to hexadecimal.

{0: '0', 1: '1', 2: '2', 3: '3', 4: '4', 5: '5', 6: '6', 7: '7', 8: '8', 9: '9', 10: 'a', 11: 'b', 12: 'c', 13: 'd', 14: 'e', 15: 'f'}


## $\textbf{Step 2.3: Implementing the Repeated Division Algorithm to convert INT to HEX}$

In [26]:
while True: # The while loop keeps the user stuck in the loop until a valid hexadecimal number is encountered.
    N = input("Enter a base 10 integer: ")
    if all([int(d) in dic_int2hex for d in N]): # checks if all the digits of n are in the int to hex dictionary
        break       
    else:
        print("Invalid integer number. Please enter a valid integer number.")
    

    
def int2hex(N) -> str:
    """
    This function converts a base_10 integer into hexadecimal number using the repititive division algorithm.
    
    Input Parameters:
    N (int)  the integer to convert.
    
    Output Parameters:
    H (str) the hexadecimal value obtained from the string.
    """
    
    # Declare a time variable to track start of algorithm implementation
    start = decimal.Decimal(time.perf_counter())
    
    result = []
    N = int(N) # convert string to integer to perform operations.
    while N%16  != 0:
        result.append(N%16) # append the remainder from each division to the result until we can no longer divide.
        N = N//16 # upate the dividend in each iteration until N == 0
    
    # Reverse the result which holds the remainder values from the earliest.
    reversed_result = result[::-1]
    
    # Convert the remainder to their mapped values using the dictionary dic_int2hex
    mapped_result = [dic_int2hex[i] for i in reversed_result]
    
    H = "".join(mapped_result) # Join back the reversed elements in the mapped_result array and save in variable H.
    
    # Declare a time variable to keep track of the end of execution.
    end = decimal.Decimal(time.perf_counter())
    time_taken = round((end - start), 10)
    
    print('\t')
    # return result and time taken to run implement the algorithm.
    return  H, time_taken


# Call the function and print the result and time taken to compute it.
hex_value, runtime_i2h = int2hex(N)

# print(int2hex.__doc__)
print(f"The equivalent hexadecimal number is {hex_value}")
print(f"The time taken is {round(runtime_i2h*10**6, 2)} mirco seconds.")

Enter a base 10 integer: 23151
	
The equivalent hexadecimal number is 5a6f
The time taken is 36.70 mirco seconds.
